storage/rocksdb/clone/donor.cc

```diff
@@ -0,0 +1,1409 @@
+
+  for (const auto &log_file_dom : wal_array) {
+    assert(log_file_dom->json_type() == enum_json_type::J_OBJECT);
+    const auto &log_file_obj = static_cast<const Json_object &>(*log_file_dom);
+
+    static const std::string path_str = "path_name";
+    const auto &path_json = *log_file_obj.get(path_str);
+    assert(path_json.json_type() == enum_json_type::J_STRING);
+    const auto &path_str_json = static_cast<const Json_string &>(path_json);
+    const auto &path = path_str_json.value();
+    assert(path[0] == '/');
+    assert(path.length() > sizeof("/x.log"));
+    assert(myrocks::has_file_extension(path, ".log"));
+    auto log_name = path.substr(1, path.length() - 1);
+
+    static const std::string size_str = "size_file_bytes";
+    const auto &size_json = *log_file_obj.get(size_str);
+    assert(size_json.json_type() == enum_json_type::J_UINT);
+    const auto &size_uint_json = static_cast<const Json_uint &>(size_json);
+    const auto size = static_cast<my_off_t>(size_uint_json.value());
+
+    donor_instance->add_log(std::move(log_name), size);
+  }
+
+  donor_instance->set_logs_added();
+}
+
+// Send the consistent clone data to the client. It comes from 1) the final
+// RocksDB checkpoint that is never rolled at this state, 2) WALs from
+// performance_schema.log_status query result.
+int rocksdb_clone_copy(handlerton *hton, THD *thd, const uchar *loc,
+                       uint loc_len, uint task_id, Ha_clone_cbk *cbk) {
+  const auto l = myrocks::clone::locator::deserialize(loc, loc_len);
+  if (l.has_applier_status())
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+  auto *const donor_instance =
+      lookup_donor_and_task_id(l, loc, loc_len, task_id);
+  if (donor_instance == nullptr)
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+
+  assert(donor_instance->get_state() ==
+         donor::donor_state::FINAL_CHECKPOINT_WITH_LOGS);
+
+  const auto err = donor_instance->handle_any_error(thd);
+  if (err != 0) return err;
+
+  cbk->set_hton(hton);
+  return donor_instance->copy(thd, task_id, *cbk);
+}
+
+// Acknowledge clone application errors that are received from the client. In
+// general, this callback is used to acknowledge the client advancing to the
+// next clone stage, but this capability is unused in MyRocks clone.
+int rocksdb_clone_ack(handlerton *, THD *thd, const uchar *loc, uint loc_len,
+                      uint task_id, int in_err, Ha_clone_cbk *) {
+  const auto l = myrocks::clone::locator::deserialize(loc, loc_len);
+  if (!l.is_valid() || l.has_applier_status())
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+  auto *const donor_instance =
+      lookup_donor_and_task_id(l, loc, loc_len, task_id);
+  if (donor_instance == nullptr)
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+
+  // ACK is only used to report errors for MyRocks clone
```

Comment:
Isn't the return code form clone_apply for `COM_RES_DATA_DESC`/`COM_RES_DATA` stored in `in_err` of the `COM_ACK` sent back to the donor, so it seems a bit surprising we don't receive `COM_ACK`s when clone_apply succeeded.

Reply:
`COM_RES_DATA_DESC`/`COM_RES_DATA` do not get an `COM_ACK` back. This is also documented in https://dev.mysql.com/worklog/task/?id=9210: "It is to be noted that during COM_EXECUTE the data is streamed one way from clone server to clone client.“

The prototype implementation did not even have `rocksdb_clone_ack` at all, it was a `nullptr` and it cloned perfectly fine if no client errors needed to be reported back to the donor.

Reply:
Ah, okay, so the `COM_ACK` are sent when an error has occurred on applying the data_desc/data. They're also sent as a response to InnoDB's DATA_DESC server responses that indicate a stage is complete.